### PR TITLE
more detailed checking of arithmetic around boundary conditions

### PR DESCRIPTION
**Pull request by ericeilertson**
This pull request introduces several improvements to error handling and arithmetic overflow protection in the firmware image processing and verification code. The changes enhance robustness by ensuring arithmetic operations are checked for overflow, updating error reporting, and strengthening unit tests for edge cases.

**Overflow protection and error handling:**

* Updated the `ImageTocEntry::overlaps` method to use `saturating_add` for computing image region ends, preventing integer overflow during address calculations.
* Improved ICCM bounds validation in `ImageVerifier` by using `checked_add` and `checked_sub` to safely compute end addresses, returning specific errors if arithmetic overflows occur. [[1]](diffhunk://#diff-52185f43b832921920e14257704d974e449b313a0c3f3bcfc0d19984fe8b0ab4L778-R786) [[2]](diffhunk://#diff-52185f43b832921920e14257704d974e449b313a0c3f3bcfc0d19984fe8b0ab4L871-R882)
* Added overflow checks when dividing image sizes by word size in `FirmwareProcessor`, returning an error if the division fails. [[1]](diffhunk://#diff-4eb62ad2fa043d17976d3f947cd8a7e1965ce0af2a340e94cc9d66e66f1e8d34L511-R514) [[2]](diffhunk://#diff-4eb62ad2fa043d17976d3f947cd8a7e1965ce0af2a340e94cc9d66e66f1e8d34L524-R530)
* Updated `copy_tbs` to use `u16::try_from` for size conversions, ensuring that TBS sizes do not overflow and reporting errors for unsupported sizes.

**Error reporting and test coverage:**

* Added a new error code and message for X509 digest conversion failures, and updated all relevant digest conversion code paths to return this error on failure. [[1]](diffhunk://#diff-009b2077ef4a38a7715447e771702e14e3f522198f0a8cc09f073a08bd02eaa8R1602-R1606) [[2]](diffhunk://#diff-c0955643e6fe9b0bdc5b84593659e0ee687e5379544f35a181e73103fcf0223eL77-R87) [[3]](diffhunk://#diff-c0955643e6fe9b0bdc5b84593659e0ee687e5379544f35a181e73103fcf0223eL108-R114) [[4]](diffhunk://#diff-c0955643e6fe9b0bdc5b84593659e0ee687e5379544f35a181e73103fcf0223eL132-R140)
* Expanded unit tests for `ImageTocEntry::overlaps` to cover edge cases near `u32::MAX`, verifying overflow protection and correct region overlap behavior.

**Code safety and correctness:**

* Added a compile-time assertion to ensure that PCR IDs used for bit shifting are within valid bounds, preventing undefined behavior.

### CODE REVIEW

File: rom/dev/src/flow/cold_reset/x509.rs [hunk 1]
```diff
@@ -74,13 +74,17 @@ impl X509 {
                 cprintln!("[idev] Sha256 KeyId Algorithm");
                 let digest = Crypto::sha256_digest(env, &data);
                 let digest: [u8; 32] = okref(&digest)?.into();
-                digest[..20].try_into().unwrap()
+                digest[..20]
```
Comment (swenson) [inline rom/dev/src/flow/cold_reset/x509.rs:77]: The checks in this file are not necessary as if unwrap() fails it will cause the ROM to panic. We ensure that no panics are in the ROM automatically, so we know that the compiler can verify that this can never occur.

File: rom/dev/src/pcr.rs [hunk 2]
```diff
@@ -48,7 +48,9 @@ impl PcrExtender<'_> {
         self.pcr_bank
             .extend_pcr(PCR_ID_FMC_JOURNEY, self.sha384, data)?;
 
-        let pcr_ids: u32 = (1 << PCR_ID_FMC_CURRENT as u8) | (1 << PCR_ID_FMC_JOURNEY as u8);
+        // Compile-time assertion that PCR IDs are valid for bit shifting
```
Comment (swenson) [inline rom/dev/src/pcr.rs:51]: I don't think this check is necessary.

File: rom/dev/src/flow/cold_reset/mod.rs [hunk 3]
```diff
@@ -109,14 +109,16 @@ pub fn copy_tbs(tbs: &[u8], tbs_type: TbsType, env: &mut RomEnv) -> CaliptraResu
     let mut persistent_data = env.persistent_data.get_mut();
     let dst = match tbs_type {
         TbsType::LdevidTbs => {
-            persistent_data.fht.ldevid_tbs_size = tbs.len() as u16;
+            persistent_data.fht.ldevid_tbs_size = u16::try_from(tbs.len())
```
Comment (swenson) [inline rom/dev/src/flow/cold_reset/mod.rs:112]: I don't think it is possible for these TBSs to ever be more than 64 KB?

File: rom/dev/src/flow/cold_reset/fw_processor.rs [hunk 4]
```diff
@@ -508,7 +508,10 @@ impl FirmwareProcessor {
 
         let fmc_dest = unsafe {
             let addr = (manifest.fmc.load_addr) as *mut u32;
-            core::slice::from_raw_parts_mut(addr, manifest.fmc.size as usize / 4)
+            let word_count = (manifest.fmc.size as usize)
```
Comment (swenson) [inline rom/dev/src/flow/cold_reset/fw_processor.rs:511]: Checks should not be necessary as we guarantee that the FMC and runtime sizes is fixed elsewhere and if it is somehow a bad image, doing a normal division should not cause an issue.

File: image/verify/src/verifier.rs [hunk 5]
```diff
@@ -775,12 +775,17 @@ impl<Env: ImageVerificationEnv> ImageVerifier<Env> {
             caliptra_cfi_lib::cfi_assert_eq_12_words(&verify_info.digest, &actual);
         }
 
-        // Overflow/underflow is checked in verify_toc
+        // Validate ICCM bounds with overflow protection
```
Comment (swenson) [inline image/verify/src/verifier.rs:778]: are these necessary? The previous comment says that these are checked in verify_toc. Is that not the case?
